multus, annotations, Consolidate interface naming scheme in a single place

RamLavi opened the pull request:
**What this PR does / why we need it**:
Since network interface names [given](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/services/multus_annotations.go#L92) in `k8s.v1.cni.cncf.io/networks` annotation by virt-controller are given by index, and later used by other places in the code, there is a need to consolidate the naming logic.
We do this by moving methods related to multus annotations to one place where the logic is kept.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

orelmisan reviewed (commented):
Thank you for the PR @RamLavi.
Is it possible to please make the change gradually over more commits?
I personally find the change too big for me to fully understand.

RamLavi commented:
> **Release note**:

see [here](https://github.com/kubevirt/kubevirt/compare/b1edd8ea45314cb78341e8c6916819d03bd31311..a25d6d8fbf65a4054f06e30a5d676788f3b0e314)

orelmisan commented:
> > Thank you for the PR @RamLavi.
> > Is it possible to please make the change gradually over more commits?
> > I personally find the change too big for me to fully understand.
> 
> @orelmisan I'ved [broken it up](https://github.com/kubevirt/kubevirt/compare/b1edd8ea45314cb78341e8c6916819d03bd31311..a25d6d8fbf65a4054f06e30a5d676788f3b0e314)

Thank you for the change, I find it helpful in reducing the cognitive load.

maiqueb commented on pkg/network/multus/annotations.go:
```diff
@@ -0,0 +1,47 @@
+	v1 "kubevirt.io/api/core/v1"
+)
+
+func IterateMultusNonDefaultNetworks(vmi *v1.VirtualMachineInstance, fn func(network v1.Network, networkRequestInterfaceName string)) {
+	multusNonDefaultNetworks := filterMultusNonDefaultNetworks(vmi.Spec.Networks)
+	for i, network := range multusNonDefaultNetworks {
+		networkRequestInterfaceName := fmt.Sprintf("net%d", i+1)
+		fn(network, networkRequestInterfaceName)
+	}
+}
```
This function is wrong imo; it is basically a poor implementation of a map function that on top of that performs extra operations - has an embedded naming scheme, plus filters the multus non-default network.

I understand why you do it; you're being cleaver - unfortunately, that costs readability, something which I would rather maximize.

Furthermore, I see that one of the callers actually rely on this function to extract data, using a side-effect. 

I would:
- explicitly have a function that exposes the interface naming scheme based on the interface index. Make that usable by whoever needs it.
- have 2 functions that receive a list of networks - you'd pass the list of already filtered networks from the caller - then let each of the properly named functions iterate the list, call the naming scheme, and do what they need to do.

orelmisan replied:
I agree with the action items @maiqueb suggested.

RamLavi replied:
I added an interfaces that does just that. 
regarding the 2 functions - not sure I fully understood, but I think the change I inserted should answer all our needs

RamLavi commented on pkg/network/multus/annotations_suite_test.go:
```diff
@@ -0,0 +1,43 @@
+			},
+		},
+	}
+}
```
moved the helper function `newMultusNetwork`.
kept the file (isn't it regarded [convention](https://onsi.github.io/ginkgo/#bootstrapping-a-suite) to keep the _suite.go file separate?)

RamLavi commented on pkg/network/multus/annotations_test.go:
```diff
@@ -0,0 +1,148 @@
+					},
+				}),
+		)
+	})
```
Moved to `It` (and also moved it away from this PR).

orelmisan reviewed (commented):
First commit message:
```
This commit creates a new Multus package for annotation related methods,
and only moves filterMultusNonDefaultNetworks function to it.
```

Can you please reason why only this function was moved and not the whole `pkg/virt-controller/services/multus_annotations.go` file?

RamLavi commented:
> First commit message:
> 
> ```
> This commit creates a new Multus package for annotation related methods,
> and only moves filterMultusNonDefaultNetworks function to it.
> ```
> 
> Can you please reason why only this function was moved and not the whole `pkg/virt-controller/services/multus_annotations.go` file?

The reason it was moved is because it was needed to be used in other places.
I haven't given thorough thought to moving the entire module to the new package. wdyt?

RamLavi commented:
/test pull-kubevirt-e2e-k8s-1.22-sig-network
/test pull-kubevirt-e2e-k8s-1.23-sig-network
/test pull-kubevirt-e2e-k8s-1.24-sig-network

maiqueb commented:
> > First commit message:
> > ```
> > This commit creates a new Multus package for annotation related methods,
> > and only moves filterMultusNonDefaultNetworks function to it.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Can you please reason why only this function was moved and not the whole `pkg/virt-controller/services/multus_annotations.go` file?
> 
> The reason it was moved is because it was needed to be used in other places. The other functions there could go into "multus/annotations.go". We should do it in a separate PR, wdyt?

I there's no reason to use those funcs from outside the virt-controller template package I see no reason to change their scope or move them around.

RamLavi commented:
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-unit-test

ormergi commented on pkg/network/podinterfacenamescheme/podinterfacenamescheme.go:
```diff
@@ -19,4 +19,37 @@
+
 const PrimaryPodInterfaceName = "eth0"
+
+type PodInterfaceNamer map[string]string
```
Why is it necessary to define new type?

RamLavi replied:
not necessary, I think this is the best approach. 
why not do a new type?

EdDev replied:
Can you define receivers to the map directly?
BTW, "namer" sounds like an interface name.

RamLavi replied:
I don't think it's possible to use a receiver on a non-named type.

>BTW, "namer" sounds like an interface name

not sure what you mean.. :)

orelmisan replied:
https://go.dev/doc/effective_go#interface-names

ormergi replied:
> not necessary, I think this is the best approach. why not do a new type?

@RamLavi if we remove the type definition there is no difference in the way it will be used, I think it doesn't improve reliability.

I understand that your intention was to make the code self-explanatory by adding methods for it that suggest
what will the naming scheme include (`mapDefaultNetworkToPodInterfaceName `mapMultusNonDefaultNetworksToPodInterfaceName`)
Note that if you get rid of the type, you can still have those methods, in just that they won't do the assignment to the map..

ormergi commented on pkg/network/podinterfacenamescheme/podinterfacenamescheme.go:
```diff
@@ -19,4 +19,37 @@
+
+type PodInterfaceNamer map[string]string
+
+func NewPodInterfaceNamer(vmi *v1.VirtualMachineInstance) PodInterfaceNamer {
```
How about `GenerateNameScheme`?

RamLavi replied:
I don't think it specific enough.

EdDev replied:
The name "pod" embedded here assumes knowledge of who is using this and for what purpose.
I would argue it should not.

As I see it, this is a network ID generator. But it does not create a generator, it generates ID/s (or if you prefer, names).
The **caller** uses it to name the pod interface, when creating the CNI annotation or when consuming the pod interfaces.

RamLavi replied:
I would argue the main purpose is for pod interface names, but  don't mind using a more general name
how about `networkNameGenerator`?

ormergi replied:
`networkNameGenerator` sounds like an object name, since this is a function I would suggest `generateNetworkName` instead.

Note that the package name suggests that it's all about pod interface names so IMO a more generic name can work too.

ormergi commented on pkg/network/podinterfacenamescheme/podinterfacenamescheme.go:
```diff
@@ -19,4 +19,37 @@
+func NewPodInterfaceNamer(vmi *v1.VirtualMachineInstance) PodInterfaceNamer {
+	podInterfaceNamer := PodInterfaceNamer{}
+
+	podInterfaceNamer.mapDefaultNetworkToPodInterfaceName(vmi)
+	podInterfaceNamer.mapMultusNonDefaultNetworksToPodInterfaceName(vmi)
```
I would inline those functions since they are not-exported and are not being reused, it's about 10 lines I think it's fine in terms of function length.
Not a must.

RamLavi replied:
I'm not thinking of lint here.
my purpose is to properly document what this map holds. 
Putting those implementation details on a function makes the code more readable.

EdDev replied:
What I do not understand is the reason for these to be implemented as receivers.

RamLavi replied:
@EdDev, if I don't use them as receivers, he I'll need to pass the map to them as an input to build the map, like in here:
```
n[multusDefaultNetwork.Name] = PrimaryPodInterfaceName
```

EdDev replied:
I may not have been clear.. I do not understand why we need an object and not direct functions to do the same.

RamLavi replied:
why not? is there a problem using that?

ormergi replied:
IMO it is redundant "overhead" because the usage stays the same at the caller whether there is a type or not,
removing the type will simplify the implementation a bit and reduce noise.

RamLavi replied:
I have to agree with you both. change it to simpler implementation

ormergi commented on pkg/virt-controller/services/multus_annotations.go:
```diff
@@ -57,11 +57,13 @@ func (mnap multusNetworkAnnotationPool) toString() (string, error) {
-	for i, network := range multusNonDefaultNetworks {
-		multusNetworkAnnotationPool.add(
-			newMultusAnnotationData(vmi, network, fmt.Sprintf("net%d", i+1)))
+	podInterfaceNamer := podinterfacenamescheme.NewPodInterfaceNamer(vmi)
```
I think the local var `podInterfaceNamer` should be renamed to `podIfacesNameScheme` or `podIfacesNames`.

RamLavi replied:
Done

ormergi reviewed (commented):
Hi Ram, a great initiative it will surely make the code more consistent.

ormergi commented:
@RamLavi what you think about using the VMI `spec.device.interfaces` names instead of the current scheme (net1, net2 ..)? 
(by specifying them in the `k8s.v1.cni.cncf.io/networks` annotation)

orelmisan commented:
> > > First commit message:
> > > ```
> > > This commit creates a new Multus package for annotation related methods,
> > > and only moves filterMultusNonDefaultNetworks function to it.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Can you please reason why only this function was moved and not the whole `pkg/virt-controller/services/multus_annotations.go` file?
> > 
> > 
> > The reason it was moved is because it was needed to be used in other places. The other functions there could go into "multus/annotations.go". We should do it in a separate PR, wdyt?
> 
> I there's no reason to use those funcs from outside the virt-controller template package I see no reason to change their scope or move them around.

But after this change we will have two packages in two different directories with roughly the same subject, thus spreading the logic to more places.
As the commit message states: `This commit creates a new Multus package for annotation related methods`, which is what `pkg/virt-controller/services/multus_annotations.go` is doing at the moment.

orelmisan commented:
> @RamLavi what you think about using the VMI `spec.device.interfaces` names instead of the current scheme (net1, net2 ..)? (by specifying them in the `k8s.v1.cni.cncf.io/networks` annotation)

@ormergi AFAIU these names are supposed to be the interface names inside the `virt-launcher` Pod.
IIRC they are limited in length.
Besides, I think it is a change that may break backwards compatibility because other code is relying on that scheme.

RamLavi commented on pkg/network/multus/multus_test.go:
```diff
@@ -0,0 +1,92 @@
+	})
+})
+
+func newMultusNetwork(name, networkName string, isDefault bool) v1.Network {
```
totally, Done

RamLavi commented on pkg/virt-controller/services/multus_annotations.go:
```diff
@@ -56,7 +58,7 @@ func (mnap multusNetworkAnnotationPool) toString() (string, error) {
 	multusNetworkAnnotationPool := multusNetworkAnnotationPool{}
 
-	multusNonDefaultNetworks := filterMultusNonDefaultNetworks(vmi.Spec.Networks)
+	multusNonDefaultNetworks := multus.FilterMultusNonDefaultNetworks(vmi.Spec.Networks)
```
makes sense.. Done

EdDev commented on pkg/network/multus/multus_test.go:
```diff
@@ -0,0 +1,100 @@
+})
+
+func createMultusSecondaryNetwork(name, networkName string) v1.Network {
+	return createMultusNetwork(name, networkName, false)
```
Why the linter did not complain, I do not know.
But you better name the boolean using a const, to make this readable.

I also do not understand why you need a wrapper if you already have a readable argument.
In fact, having a boolean as an argument is a smell, I would have avoided it and then such a wrapper will make sense.

RamLavi replied:
I think hiding the boolean in a wrapper makes it more readable

>In fact, having a boolean as an argument is a smell, I would have avoided it and then such a wrapper will make sense

Can you elaborate what do you mean? what would you suggest?

EdDev replied:
I meant this:
```
func newMultusDefaultNetwork(name, nadName) v1.Network {
    net := newMultusNetwork(name, nadName)
    net.NetworkResource.Multus.Default = true
    return net
}
```

The other one you can leave as `newMultusNetwork` or wrap with a different name (I do not see why it is needed, but it is per taste) and add nothing.

RamLavi replied:
ACK, got it, Done

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network

EdDev commented on pkg/network/multus/multus.go:
```diff
@@ -35,3 +37,12 @@ func FilterNonDefaultNetworks(networks []v1.Network) []v1.Network {
 	return multusNetworks
 }
+
+func LookUpDefaultNetwork(networks []v1.Network) (v1.Network, error) {
```
IMO this fits in `network/vmispec/network.go`.

EdDev commented on pkg/network/multus/multus.go:
```diff
@@ -35,3 +37,12 @@ func FilterNonDefaultNetworks(networks []v1.Network) []v1.Network {
+			return network, nil
+		}
+	}
+	return v1.Network{}, fmt.Errorf("no multus default network found")
```
Existing patterns in `vmispec` are returning a `nil` if nothing was found, which seems a better fit IMO.
It would be even better if the lookup will return the actual network (original one) and not a copy.

If someone needs to expose the failure to find a network, then that caller should raise an error.

RamLavi replied:
I was struggling with myself on that, and finally decided to have it return a non pointer for unit-test reasons.. I have no problem to change it back..

EdDev replied:
If it returns the original network, then the caller can mutate it in-place if it needs to. If it returns a copy, it needs to override the original one which I find less attractive.

In tests you may even check identity instead of equality.

RamLavi replied:
yep I agree.

EdDev commented on pkg/network/multus/multus_test.go:
```diff
@@ -68,6 +68,53 @@ var _ = Describe("Multus", func() {
 		)
 	})
+	Context("LookUpDefaultNetwork", func() {
+		type lookUpDefaultNetworkParams struct {
+			inputNetworks []v1.Network
+			expectNetwork v1.Network
+		}
```
Same questions as before... why do we need a struct and not two arguments as input to the test table.

RamLavi replied:
It boils down to personal style, but I think this way has the benefit of separating the input "params" and the internal ones.
Do you think this way has bad traits?

EdDev replied:
I am having hard time to understand it. I mean, the argument is called `params` which is indeed the function inputs, but in a structured manner instead of individual arguments.
If the member names are defined as the function arguments, the meaning and clarity is not degraded IMO.

> separating the input "params" and the internal ones

Could you please clarify what are these "internal ones"?

I am not sure, but are we trying here to find arguments on why we should use a single struct as an argument instead of multiple variables? `f(a, b, c)` ==>> `f(s)`
This seems to me like a general question, not limited to tests.

I just do not see the added value at the moment, only drawbacks in trying to understand what this struct is and why it was needed. We are adding here ` lookUpDefaultNetworkParams{` for each call and the definition of the struct once.

RamLavi replied:
>Could you please clarify what are these "internal ones"?

By that I mean variables that are created inside the test, as oppose to ones that are given to it.

>I am not sure, but are we trying here to find arguments on why we should use a single struct as an argument instead of multiple variables? f(a, b, c) ==>> f(s)
This seems to me like a general question, not limited to tests.

I wouldn't go as far as generalizing it to a general rule, but generally speaking, in tests (and specifically DescribeTable tests) I prefer using a struct if there is more than 1 params to insert. 
If you don't have a significant drawback then like I initially said - it boils down to style...

RamLavi commented on pkg/network/multus/multus_test.go:
```diff
@@ -0,0 +1,100 @@
+
+var _ = Describe("Multus", func() {
+	Context("FilterNonDefaultNetworks", func() {
+		type filterNonDefaultNetworksParams struct {
+			inputNetworks          []v1.Network
+			expectFilteredNetworks []v1.Network
+		}
```
I find it neater to use a params struct if there is more than one object..

EdDev replied:
I'm not sure I understand what is "neater" about it. From me, it adds overhead which is not needed.
I think it fits the classic table testing that comes with Go, but this describe-table is not that.

ormergi commented:
> @ormergi AFAIU these names are supposed to be the interface names inside the `virt-launcher` Pod. IIRC they are limited in length. 

Correct, the CNI will set the interfaces with these names (net1, net2..), the limitation is actually a limitation of the Linux kernel (15 character max IIRC).
I am not saying it won't require a change to Kubevirt admission webhook (on the VMI interfaces length)..

> Besides, I think it is a change that may break backwards compatibility because other code is relying on that scheme.
How is it different from introducing any other feature that requires changing more than one component?

RamLavi commented:
> > @ormergi AFAIU these names are supposed to be the interface names inside the `virt-launcher` Pod. IIRC they are limited in length.
> 
> Correct, the CNI will set the interfaces with these names (net1, net2..), the limitation is actually a limitation of the Linux kernel (15 character max IIRC). I am not saying it won't require a change to Kubevirt admission webhook (on the VMI interfaces length)..
> 
> > Besides, I think it is a change that may break backwards compatibility because other code is relying on that scheme.
> 
> How is it different from introducing any other feature that requires changing more than one component?

I think it would make things easier, but I don't think it should be done as part of the design. This kind of change deserves a design of its own.

RamLavi commented on pkg/network/multus/multus.go:
```diff
@@ -0,0 +1,37 @@
+	v1 "kubevirt.io/api/core/v1"
+)
+
+func FilterNonDefaultNetworks(networks []v1.Network) []v1.Network {
```
Done

RamLavi commented on pkg/network/multus/multus_test.go:
```diff
@@ -0,0 +1,100 @@
+				},
+				expectFilteredNetworks: nil,
+			}),
+			Entry("when there is a default Multus network", filterNonDefaultNetworksParams{
```
Done

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network

EdDev commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go:
```diff
@@ -26,13 +26,14 @@ import (
 
 func CreateHostDevices(vmi *v1.VirtualMachineInstance) ([]api.HostDevice, error) {
-	SRIOVInterfaces := filterVMISRIOVInterfaces(vmi)
+	SRIOVInterfaces := vmispec.FilterSRIOVInterfaces(vmi.Spec.Domain.Devices.Interfaces)
```
Can you use a different commit for this change?

RamLavi replied:
why? it's holds under the title of moving functions to a different package.

RamLavi replied:
moved to commit

EdDev reviewed (changes requested):
> vmispec, Move functions to package 

How about:
network, vmispec: Move the multus network filtering helper

I would also not use wording like "this commit ...". Just mention directly what it does.

EdDev commented on pkg/network/networknamescheme/networknamescheme.go:
```diff
@@ -0,0 +1,22 @@
+ *
+ */
+
+package networknamescheme
```
We are already under the `network` package, so just using `namescheme` may fit better.
I do see a problem for the long run with this new package... it may become overloaded with all kind of unrelated names.

But I do not know if I have a better place for this naming constants.
The closes I can think of is under `network.link`, so it will be `network.link.namescheme`. WDYT?
I must admit I am not really sure my suggestion fits.

RamLavi replied:
I agree, but I don't think your suggestion makes sense I'm afraid..

Changed to `namescheme`

EdDev commented on pkg/network/networknamescheme/networknamescheme.go:
```diff
@@ -19,4 +19,35 @@
+
 const DefaultNetworkName = "eth0"
+
+func GetNetworkNameSchemeMap(vmi *v1.VirtualMachineInstance) map[string]string {
```
How about `CreateNetworkToLinkScheme`?
Also, pleas pass in only the information needed to create the data (i.e. `vmi.Spec.Networks`).

RamLavi replied:
the naming issue of networks/interfaces is indeed problematic but I don't think links should be into this.. we have already too many names for "network".

EdDev commented on pkg/network/networknamescheme/networknamescheme.go:
```diff
@@ -19,4 +19,35 @@
+	return networkNameSchemeMap
+}
+
+func mapMultusNonDefaultNetworksToPodInterfaceName(vmi *v1.VirtualMachineInstance, networkNameSchemeMap map[string]string) {
```
I think you can create and return an initialized and populated map instead of mutating an existing one.
Whenever we mutate, we should feel bad about it :)

maiqueb replied:
no mutatos.

RamLavi replied:
I agree :)

EdDev commented on pkg/network/networknamescheme/networknamescheme.go:
```diff
@@ -19,4 +19,35 @@
+	}
+}
+
+func mapDefaultNetworkToPodInterfaceName(vmi *v1.VirtualMachineInstance, networkNameSchemeMap map[string]string) {
```
Same comment here, I would try to avoid mutation like this.
But for that, in this case, it will need to be done at the caller.
```
if multusDefaultNetwork, err := vmispec.LookUpDefaultNetwork(vmi.Spec.Networks); err == nil {
    networkNameSchemeMap[multusDefaultNetwork.Name] = DefaultNetworkName
}
```

An alternative is to return the mutated map.

maiqueb replied:
plus one on avoiding mutating the inner map.

RamLavi replied:
+1

EdDev commented on pkg/network/vmispec/network.go:
```diff
@@ -43,3 +47,12 @@ func FilterMultusNonDefaultNetworks(networks []v1.Network) []v1.Network {
+func LookUpDefaultNetwork(networks []v1.Network) (*v1.Network, error) {
+	for _, network := range networks {
+		if network.Multus == nil || network.Multus.Default {
+			return &network, nil
```
It is discouraged to pass the address of the value which comes from the range.
You better return the original entity address: `&networks[i]`.

RamLavi replied:
totally! ! !

RamLavi commented on pkg/network/vmispec/network_test.go:
```diff
@@ -0,0 +1,82 @@
+package vmispec_test
```
Done

RamLavi commented on pkg/network/vmispec/network_test.go:
```diff
@@ -0,0 +1,82 @@
+)
+
+var _ = Describe("Network", func() {
+	Context("FilterMultusNonDefaultNetworks", func() {
```
Done

RamLavi commented on pkg/network/vmispec/network_test.go:
```diff
@@ -0,0 +1,82 @@
+			inputNetworks          []v1.Network
+			expectFilteredNetworks []v1.Network
+		}
+		DescribeTable("should return the only non default networks", func(params FilterMultusNonDefaultNetworksParams) {
```
>How about: should return only secondary networks.

Like!

> The DescribeTable ability to handle dynamic func signature can be used, removing the need to do this boilerplate structure.

I don't think it's a must, but I don't understand your string objection to it.

maiqueb reviewed (changes requested):
Just pushing my preliminary review before this changes too much.

maiqueb commented on pkg/network/setup/podnic.go:
```diff
@@ -107,9 +109,9 @@ func newPodNIC(vmi *v1.VirtualMachineInstance, network *v1.Network, handler netd
-	podInterfaceName, err := composePodInterfaceName(vmi, network)
-	if err != nil {
-		return nil, err
+	podInterfaceName, exists := podinterfacenamescheme.NewPodInterfaceNamer(vmi)[network.Name]
```
By clunky I mean two things:
- you use the map API inconsistently - I think you should either **always** handle the second output variable to know if it exists **or** never use it at all. I favor the second.
- you pass the list of networks on the VMI, from which your new struct builds a map. Then you ask it about the network you know is there. I think the API should somehow reflect the fact the network **will always** be there - or something **really** wrong occurred. I'd rather see the API return `string, error` or just `string` and panic when the network is not found. AFAIU, we shouldn't ever reach that scenario.

RamLavi replied:
removed the PodInterfaceNamer object in favor of a simple function.
Now it returns a map, so I think searching over the map and returning error if !exists is good enough for what we want. wdyt?

maiqueb commented on pkg/network/networknamescheme/networknamescheme.go:
```diff
@@ -19,4 +19,35 @@
+func GetNetworkNameSchemeMap(vmi *v1.VirtualMachineInstance) map[string]string {
+	networkNameSchemeMap := map[string]string{}
+
+	mapDefaultNetworkToPodInterfaceName(vmi, networkNameSchemeMap)
+	mapMultusNonDefaultNetworksToPodInterfaceName(vmi, networkNameSchemeMap)
```
I would prefer to see something where these helpers return the map rather than passing one via input arg then mutate it.

RamLavi commented on pkg/network/podinterfacenamescheme/podinterfacenamescheme.go:
```diff
@@ -0,0 +1,55 @@
+
+type PodInterfaceNamer map[string]string
+
+func NewPodInterfaceNamer(vmi *v1.VirtualMachineInstance) PodInterfaceNamer {
```
the motivation is that we want to make sure the the networks we are naming are the `vmi.spec.networks` ones..
The only way this naming can be consistent is if we give them the same network list every time. 
The way I thought to impose it is to pass the entire vmi... 
But I'm open to suggestions of course :)

RamLavi replied:
changed. Thanks!

RamLavi commented on pkg/network/setup/podnic.go:
```diff
@@ -23,6 +23,8 @@ import (
 	"os"
 
+	"kubevirt.io/kubevirt/pkg/network/podinterfacenamescheme"
+
```
no reason. fixed

RamLavi commented on pkg/network/vmispec/network.go:
```diff
@@ -30,3 +30,16 @@ func LookupPodNetwork(networks []v1.Network) *v1.Network {
+			if network.Multus.Default {
+				continue
+			}
+			multusNetworks = append(multusNetworks, network)
```
moved to function `isSecondaryMultusNetwork`

RamLavi commented on pkg/network/podinterfacenamescheme/podinterfacenamescheme.go:
```diff
@@ -0,0 +1,55 @@
+func (n PodInterfaceNamer) mapMultusNonDefaultNetworksToPodInterfaceName(vmi *v1.VirtualMachineInstance) {
+	multusNonDefaultNetworkList := multus.FilterNonDefaultNetworks(vmi.Spec.Networks)
+	for i, network := range multusNonDefaultNetworkList {
+		n[network.Name] = fmt.Sprintf("net%d", i+1)
```
Done

RamLavi commented:
addressed comments [here](https://github.com/kubevirt/kubevirt/compare/c9dcf262837cbe107b2953314bcda95d2f0c9d1c..5475f3469946ce865d10063f3934abd222597a08)

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network

ormergi commented on pkg/network/namescheme/networknamescheme_test.go:
```diff
@@ -20,11 +20,75 @@
+	})
+})
+
+func newMultusNetwork(name, networkName string, isDefault bool) virtv1.Network {
```
Could you please apply the same pattern from previous commit here? (avoid passing a boolean to function)

RamLavi replied:
Done

ormergi commented on pkg/network/vmispec/network_test.go:
```diff
@@ -55,6 +55,39 @@ var _ = Describe("Network", func() {
+		Expect(err).To(HaveOccurred())
+		Expect(defaultNetwork).To(Equal(expectNetwork))
+	},
+		Entry("when there are no networks", []v1.Network{}, nil),
```
Does it make sense to add a scenario where `[]v1.Network` is `nil` (similar to CreateNetworkNameScheme test suite)?

EdDev replied:
I think it is redundant to have two, this is a language feature (to behave similar to `nil` and an empty list.
But if the author wants to be pedantic, a `nil` may represent a more realistic usage.
IMO it is a nit.

ormergi commented on pkg/network/namescheme/networknamescheme.go:
```diff
@@ -19,4 +19,35 @@
+
 const PrimaryPodInterfaceName = "eth0"
+
+func CreateNetworkNameScheme(vmi *v1.VirtualMachineInstance) map[string]string {
```
Please consider changingg the func signature to receive []v1.Network, its seems that its the only thing that being used.

nit: I would call this func `Generate..` instead of `Create..` because it could emphasize that the network-name-scheme is not created out of anything.
There is an algorithm that processes the data we have (vmi.spec.networks) and produces something out of it. 
Not a must.

RamLavi replied:
>Please consider changing the func signature to receive []v1.Network, it seems that it's the only thing that is used. 

Done

>nit: I would call this func Generate..

I think create and generate have the same meaning in this scope.

ormergi commented on pkg/virt-controller/services/multus_annotations.go:
```diff
@@ -58,10 +58,13 @@ func (mnap multusNetworkAnnotationPool) toString() (string, error) {
+	netowrkNameScheme := namescheme.CreateNetworkNameScheme(vmi)
+	for _, network := range vmi.Spec.Networks {
+		podInterfaceName := netowrkNameScheme[network.Name]
+		if podInterfaceName != namescheme.PrimaryPodInterfaceName {
```
In the scenario where there is a VMI with a secondary network (of type Multus) configured as the default (default=true) network,
and have the exact same name as `namescheme.PrimaryPodInterfaceName` (eth0),
does it make sense to filter it out?

EdDev replied:
> In the scenario where there is a VMI with a secondary network (of type Multus) configured as the default (default=true) network

There is not such thing. When it is marked "default", it means it is the primary network.

> and have the exact same name as `namescheme.PrimaryPodInterfaceName` (eth0), does it make sense to filter it out?

If both are defined (pod and multus-default), then it is a configuration error.

ormergi replied:
What I meant is a VMI with Multus network and its name is eth0, at first, I thought it will filter it out 
but I got it wrong. Sorry for the noise.

ormergi commented on pkg/network/vmispec/network_test.go:
```diff
@@ -0,0 +1,88 @@
+
+		Expect(filteredNetworks).To(Equal(expectFilteredNetworks))
+	},
+		Entry("when there are no networks", []v1.Network{}, nil),
```
Does it make sense to add a scenrio where `[]v1.Network` is `nil`? (similar to CreateNetworkNameScheme tests suite)

EdDev replied:
It does not matter if it is `nil` or an empty initialized list.
Seems redundant to me. You could argue that `nil` is a better fit, because in reality this is what one would have expected to see here and not an empty list.
But I think it is a nit.

ormergi reviewed (commented):
Thanks for the changes!

RamLavi commented:
@ormergi fixed [here](https://github.com/kubevirt/kubevirt/compare/9d48d9d20cfeee682b12e2477d597754d6ce49be..7ab923dc8e95b440e9ac3561db55119e35eb5b24)
Thanks! :)

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network

EdDev reviewed (approved):
Thank you!

ormergi reviewed (approved):
Thank you, good work!

orelmisan reviewed (approved):
Thank you @RamLavi!
Please note the PR is still marked as a draft.

maiqueb commented on pkg/network/namescheme/networknamescheme.go:
```diff
@@ -0,0 +1,56 @@
+	return networkNameSchemeMap
+}
+
+func secondaryNetworkName(idx int) string {
```
this is not a network name, it is an interface name.

RamLavi replied:
Done

maiqueb commented on pkg/network/namescheme/networknamescheme_suite_test.go:
```diff
@@ -13,22 +13,18 @@
  * limitations under the License.
  *
- * Copyright 2021 Red Hat, Inc.
+ * Copyright 2022 Red Hat, Inc.
```
All changes to this file look very weird to me.

RamLavi replied:
I think I switched between `networknamescheme_test.go` and `networknamescheme_suite_test.go`, oops! will fix

RamLavi replied:
Done

maiqueb commented on pkg/network/namescheme/networknamescheme_test.go:
```diff
@@ -0,0 +1,90 @@
+				map[string]string{
+					"default": namescheme.PrimaryPodInterfaceName,
+				}),
+			Entry("when default multus networks exist",
```
would it make sense to test a scenario with a multus default network + pod network ? 

Does our admitter reject that ?

RamLavi replied:
that's a non-valid configuration. it should error, but not by the networknameshceme

maiqueb replied:
Please just make sure the admitter rejects it - I really think it should.

Just saying that unless the admitter blocks it, it may reach here.

RamLavi replied:
checking.. will update on this thread

RamLavi replied:
Indeed the admitter is [blocking](https://github.com/kubevirt/kubevirt/blob/5f4d9957608c4ae3d8fd1fe40a9a70f58da50408/pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go#L176-L178) the config you mention

maiqueb commented on pkg/network/vmispec/network_test.go:
```diff
@@ -0,0 +1,115 @@
+		Expect(filteredNetworks).To(Equal(expectFilteredNetworks))
+	},
+		Entry("when there are no networks", []v1.Network{}, nil),
+		Entry("when there are no non-default networks", []v1.Network{podNetwork}, nil),
```
"no non-default" broke me. Plus, the thing here - imo - is not the networks are "non-default" - the networks are not even from multus. 

Could we get rid of the double negative and just say something like "when there is only the pod network".

RamLavi replied:
yep, thanks!

maiqueb commented on pkg/network/vmispec/network_test.go:
```diff
@@ -0,0 +1,115 @@
+	multusSecondaryNetwork1 := createMultusSecondaryNetwork("network1", "default/nad1")
+	multusSecondaryNetwork2 := createMultusSecondaryNetwork("network2", "default/nad2")
+	multusSecondaryNetwork3 := createMultusSecondaryNetwork("network3", "default/nad3")
+	DescribeTable("should return the only non default networks", func(inputNetworks, expectFilteredNetworks []v1.Network) {
```
I miss the word `multus` in the description - i,e, `multus non-default networks`

RamLavi replied:
Done

maiqueb commented on pkg/network/vmispec/vmispec_suite_test.go:
```diff
@@ -1,3 +1,22 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
```
what's the point in this ?

RamLavi replied:
this file was simply missing  license signature, so I added it.

maiqueb replied:
but you don't touch anything else in the file, it's surprising to see this change here.

(not saying to drop it).

RamLavi replied:
yeah it's a little piggy back I added, If you'll insist I'll add it in a separate PR.

maiqueb replied:
Feel free to keep it.

maiqueb commented on pkg/virt-controller/services/multus_annotations.go:
```diff
@@ -56,10 +58,13 @@ func (mnap multusNetworkAnnotationPool) toString() (string, error) {
-	for i, network := range multusNonDefaultNetworks {
-		multusNetworkAnnotationPool.add(
-			newMultusAnnotationData(vmi, network, fmt.Sprintf("net%d", i+1)))
+	networkNameScheme := namescheme.CreateNetworkNameScheme(vmi.Spec.Networks)
+	for _, network := range vmi.Spec.Networks {
+		podInterfaceName := networkNameScheme[network.Name]
+		if podInterfaceName != "" && podInterfaceName != namescheme.PrimaryPodInterfaceName {
+			multusNetworkAnnotationPool.add(
+				newMultusAnnotationData(vmi, network, podInterfaceName))
+		}
```
I have to say I find the previous code a lot simpler.

RamLavi replied:
introduced a function, hope it help readability

maiqueb replied:
It's better, but I still think it looks worse - at least, it looked simpler before.

My overall question is since you're trying to unify the interface naming scheme, why do all this ? Couldn't you just have exported a function that names the interfaces (you're doing it), and use that anywhere a pod interface name is expected ? 

What's the advantage of adding this naming scheme type ? afaiu, I can also choose to not use your new type and compute the interface name whatever way I want. 

From my perspective, achieving consistency in the iface naming scheme can be achieved in simpler way, and the real value this PR is adding is increasing the test coverage of the helper function that filter / look for particular interface types from a list of networks.

RamLavi replied:
>My overall question is since you're trying to unify the interface naming scheme, why do all this ? Couldn't you just have exported a function that names the interfaces (you're doing it), and use that anywhere a pod interface name is expected ?
What's the advantage of adding this naming scheme type ? afaiu, I can also choose to not use your new type and compute the interface name whatever way I want.

Not sure what you mean by "adding this naming scheme type" - I changed that a few pushes back (as indeed it was an overkill) 
What this PR does now is using a function `CreateNetworkNameScheme` that returns a simple map (eg {"network1": "net1", "network2": "net2"} etc...) - this map is used wherever it is needed, like podnic.go and multus-annotations.go (and later on the SRIOV remapping fix PR).
So I don't understand your concern - can you elaborate?

maiqueb replied:
My concern is the code looks more complex now than it did before.

Trying to re-phrase my question; couldn't you just have exported a function that names the interfaces (you're doing it), and use that anywhere a pod interface name is expected ?

What's the advantage of using this approach ?

Anyway, seems you've convinced a maintainer and more people this change is good for the project. I won't fight to overthrow it.

RamLavi replied:
>Trying to re-phrase my question; couldn't you just have exported a function that names the interfaces (you're doing it), and use that anywhere a pod interface name is expected ?
What's the advantage of using this approach ?

That sounds similar to what I initially did with the function with the closure input. Or maybe I didn't get to the bottom of what you're saying :).
I think returning a map that can be used multiple times instead of iterating every time over the networks is an advantage. 

>Anyway, seems you've convinced a maintainer and more people this change is good for the project. I won't fight to overthrow it.

ACK, thanks for your input, it was important for me to understand if you have crucial concerns about this.
Seeing you don't - I'll unhold.

RamLavi commented:
@maiqueb fixed [here](https://github.com/kubevirt/kubevirt/compare/afe864565b4d99289a540b66989fe5fbd5e46ffa..2ff54c618ffe3323b1c67dce378ade65535a4b2b)

RamLavi commented:
/retest-required

AlonaKaplan commented:
/approve

RamLavi commented:
/hold
Let's wait for @maiqueb to say if it's good with him

RamLavi commented:
/retest-required

RamLavi commented:
/unhold

RamLavi commented:
/retest-required

RamLavi commented:
/cherry-pick release-0.56

RamLavi commented:
failed on release-0.56, but  doesn't hurt to try on release-0.55 :)  
/cherry-pick release-0.55
